### PR TITLE
Feature/stopping powers

### PR DESCRIPTION
**Pull request by david8dixon**
Implementation of DEDX1 interpretation exposing only the standard stopping power model.

### CODE REVIEW

**Comment by david8dixon** [comment]
I need to add some assertions to protect against querying outside of the density or temperature grid, but otherwise the CRTP is 95% or more done.

File: src/ACEtk/interpretation/DEDX1/CRTP.hpp [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+  auto numEnergies() const {return this->table.get().data.NXS( 4 );}
+  auto numDensities() const {return this->table.get().data.NXS( 5 );}
+  auto numTemperatures() const {return this->table.get().data.NXS( 6 );}
+  auto gridStart() const {return this->table.get().data.JXS( 2 );}    
+  
+public:
+  CRTP( const Table& table ) : table(table) {}
+
+  auto logEnergies() const {
+    const auto length = this->numEnergies();
```
Comment (jlconlin) [inline src/ACEtk/interpretation/DEDX1/CRTP.hpp:26]: Please move these functions to a separate file. The style guide states that every function that is more than one line long needs to be in it's own file. The files should be in the `src/` directory and have the same name as the function.

File: src/ACEtk/interpretation/DEDX1/CRTP.hpp [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+
+protected:
+
+  template<typename Range, typename T>
+  auto lowerBoundIndex(const Range& range, const T value) const noexcept(false) {
+    auto first = ranges::front(range);
+    auto last = ranges::back(range);
+    
+    if ( value < first ) {
+      throw std::domain_error("Queried value is below lowest value in table");
```
Comment (jlconlin) [inline src/ACEtk/interpretation/DEDX1/CRTP.hpp]: Please log an error (`Log::error(...)`) before throwing an exception.

**Review by jlconlin — DISMISSED** [review]

File: src/ACEtk/interpretation/DEDX1.hpp [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+
+   #include "ACEtk/interpretation/DEDX1/CRTP.hpp" 
+  
+public:
+
+  DEDX1( Table&& table) : table(std::move(table)){}
+
+  auto atomicNumber() const { return Z( this->table ); }
+  auto energySpacing() const { return energySpacing( this->table ); }
+  auto densitySpacing() const { return densitySpacing( this->table ); }
```
Comment (apmccartney) [inline src/ACEtk/interpretation/DEDX1.hpp:19]: How might a client consume the energySpacing and densitySpacing methods?

File: src/ACEtk/interpretation/DEDX1/CRTP.hpp [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+
+    return this->makeS0(this->numDensities(), std::move(stoppingPowerRange));
+
+  }
+
+};
+
+/*
+
+  auto energies() const {
```
Comment (apmccartney) [inline src/ACEtk/interpretation/DEDX1/CRTP.hpp:31]: I would prefer these methods be included in the API

File: src/ACEtk/interpretation/DEDX1/CRTP.hpp [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+      const auto start = self->table.get().data.JXS( index_to_start );
+
+      return self->table.get().data.XSS( start, length )
+             | ranges::view::chunk(self->numEnergies());
+    }();
+
+    using LogEnergies = decltype(this->logEnergies());    
+    using LogValues = std::decay_t<decltype(logStoppingPowerRanges.front())>;
+
+    struct StoppingPower {
```
Comment (apmccartney) [inline src/ACEtk/interpretation/DEDX1/CRTP.hpp]: please protect the fields

File: src/ACEtk/interpretation/DEDX1/CRTP.hpp [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+      double logTemperature_;
+      double logDensity_;
+      LogEnergies logEnergies_;
+      LogValues logValues_;
+      auto logTemperature() const {return this->logTemperature_;}
```
Comment (apmccartney) [inline src/ACEtk/interpretation/DEDX1/CRTP.hpp]: please add non-log getters

File: src/ACEtk/interpretation/DEDX1/S0.hpp [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+    : Range(std::forward<Args>(args)...), nDensities(nDensities) {}
+
+  auto floor(DenT density) const {
+    auto value = std::log(density.value);
+    auto end_it = ranges::next(this->begin(), nDensities);
+    auto it =
+      ranges::lower_bound(this->begin(), end_it, value, std::less<>{},
+			  &StoppingPower::logDensity);
+    if ( it == end_it ){ throw std::domain_error("Could not find density in range"); }
+    const bool match = (*it).logDensity_ == value;
```
Comment (apmccartney) [inline src/ACEtk/interpretation/DEDX1/S0.hpp]: use getter

File: src/ACEtk/interpretation/DEDX1/S0.hpp [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+    return S1<decltype(result), TempT, Projection>{logTemperature, std::move(result)};
+  }
+
+  auto floor(TempT temperature) const {
+    auto value = std::log(temperature.value);
+    auto it = ranges::lower_bound(this->begin(), this->end(), value, std::less<>{},
+				  &StoppingPower::logTemperature);
+    if ( it == this->end() ){ throw std::domain_error("Could not find temperature in range"); }
+    int distance = ranges::distance(this->begin(), it);
+    if( (*it).logTemperature_ != value) { distance -= nDensities; }
```
Comment (apmccartney) [inline src/ACEtk/interpretation/DEDX1/S0.hpp]: please use getter

File: src/ACEtk/interpretation/DEDX1/test/dedx1.test.cpp [hunk 9]
```diff
@@ -0,0 +1,260 @@
...
+      auto sp = standard.stoppingPowers();
+      auto base = std::vector<double>(2,0.0);
+      double i = 0.0;
+      auto increment = [&](auto&& val){return val+i;};	
+
+      AND_THEN("iterating through the stopping power values and "
+	       "querying for the values"){
+
+	RANGES_FOR(auto&& entry, sp){
+	  auto reference = base | ranges::view::transform(increment);
```
Comment (apmccartney) [inline src/ACEtk/interpretation/DEDX1/test/dedx1.test.cpp]: the intent is that this is a repeated value, right? prefer `ranges::view::repeat_n`

File: src/ACEtk/interpretation/DEDX1/test/dedx1.test.cpp [hunk 10]
```diff
@@ -0,0 +1,260 @@
...
+  }
+
+  auto standard = dedx1.standardWithoutCutoff();  
+  GIVEN("a standard model without a cutoff"){
+    auto s0 = standard.stoppingPowers();
+
+      WHEN("Querying via the parameterized interface"){
+
+      auto hh = s0.ceil( 1.5 * mev ).ceil( 1.5e24 / cc );
+      REQUIRE( hh.logDensity_ == 55.85901 );
```
Comment (apmccartney) [inline src/ACEtk/interpretation/DEDX1/test/dedx1.test.cpp]: Please update to use getters

**Review by apmccartney — CHANGES_REQUESTED** [review]
See comments

File: src/ACEtk/interpretation/DEDX1/CRTP.hpp (hunk 1 above)
Comment (david8dixon) [reply src/ACEtk/interpretation/DEDX1/CRTP.hpp:26]: I will create a few files for some of the accessors in efforts to keep the class < 100 lines.

File: src/ACEtk/interpretation/DEDX1/CRTP/src/S0.hpp [hunk 11]
```diff
@@ -0,0 +1,53 @@
+template<typename Range>
+class S0 : public Range {
+
+  int nDensities;
+  DenT min_den;
```
Comment (apmccartney) [inline src/ACEtk/interpretation/DEDX1/CRTP/src/S0.hpp]: Can these fields be removed as we discussed or did they prove useful?

File: src/ACEtk/interpretation/DEDX1/CRTP/src/S0.hpp [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+  DenT min_den; 
+  DenT max_den;
+  TempT min_temp;
+  TempT max_temp;
+  
+protected:
+  using StoppingPower =
+    std::decay_t<decltype(std::declval<const Range&>().front())>;
+
+  bool out_of_bounds(TempT value) const {
```
Comment (apmccartney) [inline src/ACEtk/interpretation/DEDX1/CRTP/src/S0.hpp]: prefer camelCase in NJOY work, but not public so less of an issue

**Review by apmccartney — APPROVED** [review]

**Pull request merged**